<a href="https://colab.research.google.com/github/FabriciaDev/data_Science_alura_oracle/blob/main/Python_para_Data_Science_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <font color=green> PYTHON PARA DATA SCIENCE - PANDAS
---

# <font color=green> 1. INTRODUÇÃO AO PYTHON
---

# 1.1 Introdução

> Python é uma linguagem de programação de alto nível com suporte a múltiplos paradigmas de programação. É um projeto *open source* e desde seu surgimento, em 1991, vem se tornando uma das linguagens de programação interpretadas mais populares.
>
> Nos últimos anos Python desenvolveu uma comunidade ativa de processamento científico e análise de dados e vem se destacando como uma das linguagens mais relevantes quando o assundo é ciência de dados e machine learning, tanto no ambiente acadêmico como também no mercado.

# 1.2 Instalação e ambiente de desenvolvimento

### Instalação Local

### https://www.python.org/downloads/
### ou
### https://www.anaconda.com/distribution/

### Google Colaboratory

### https://colab.research.google.com

### Verificando versão

In [ ]:
!python -V

Python 3.10.12


# 1.3 Trabalhando com dados

Antes de mais nada é preciso importar o arquivo db.csv que vai ser usado indo até a aba de arquivos e fazer upload do arquivo.

In [ ]:
import pandas as pd

*Ao executarmos receberemos alguns erros, pois existe um argumento faltando: sep, que define o separador utilizado no conjunto de dados. Nesse caso, o separador é ponto e vírgula (;). Poderemos então visualizar nosso conjunto a partir da variável dataset.*

In [ ]:
dataset = pd.read_csv('db.csv', sep = ';')
dataset

Nome                  Motor   Ano  Quilometragem  \
0            Jetta Variant        Motor 4.0 Turbo  2003        44410.0   
1                   Passat           Motor Diesel  1991         5712.0   
2                 Crossfox        Motor Diesel V8  1990        37123.0   
3                      DS5        Motor 2.4 Turbo  2019            NaN   
4         Aston Martin DB4        Motor 2.4 Turbo  2006        25757.0   
..                     ...                    ...   ...            ...   
253           Phantom 2013               Motor V8  2014        27505.0   
254  Cadillac Ciel concept               Motor V8  1991        29981.0   
255             Classe GLK  Motor 5.0 V8 Bi-Turbo  2002        52637.0   
256       Aston Martin DB5           Motor Diesel  1996         7685.0   
257                  Macan        Motor Diesel V6  1992        50188.0   

     Zero_km                                         Acessórios      Valor  
0      False  ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
1      False  ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
2      False  ['Piloto automático', 'Controle de estabilidad...   72832.16  
3       True  ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
4      False  ['Rodas de liga', '4 X 4', 'Central multimídia...   92612.10  
..       ...                                                ...        ...  
253    False  ['Controle de estabilidade', 'Piloto automátic...   51759.58  
254    False  ['Bancos de couro', 'Painel digital', 'Sensor ...   51667.06  
255    False  ['Rodas de liga', 'Controle de tração', 'Câmbi...   68934.03  
256    False  ['Ar condicionado', '4 X 4', 'Câmbio automátic...  122110.90  
257    False  ['Central multimídia', 'Teto panorâmico', 'Vid...   90381.47  

[258 rows x 7 columns]

*Repare que nosso retorno só nos mostrou dez linhas, mas é possível visualizar outras. Utilizando a função pd.set_option() é possível definir alguns parâmetros do dataframe, como o número máximo de linhas retornadas, algo que configuraremos com display.max_rows seguido do número determinado de linhas, neste exemplo 1000.*

In [ ]:
import pandas as pd
pd.set_option('display.max_rows', 1000)

*Se executarmos dataset novamente, receberemos todas as 258 linhas - não foram mostradas 1000 pois nosso conjunto não possui esse total de informações. Também temos uma opção para exibir mais colunas, chamada display.max_columns. Como gostaríamos de manter a configuração padrão, comentaremos ambas as linhas.*

In [ ]:
import pandas as pd
pd.set_option('display.max_rows', 10000)
# pd.set_option('display.max_columns', 1000)

dataset

Nome                  Motor   Ano  \
0                             Jetta Variant        Motor 4.0 Turbo  2003   
1                                    Passat           Motor Diesel  1991   
2                                  Crossfox        Motor Diesel V8  1990   
3                                       DS5        Motor 2.4 Turbo  2019   
4                          Aston Martin DB4        Motor 2.4 Turbo  2006   
5                             Palio Weekend          Motor 1.8 16v  2012   
6                                        A5        Motor 4.0 Turbo  2019   
7                            Série 3 Cabrio           Motor 1.0 8v  2009   
8                              Dodge Jorney          Motor 3.0 32v  2010   
9                                    Carens  Motor 5.0 V8 Bi-Turbo  2011   
10                                Ford Edge        Motor Diesel V6  2002   
11                                  Courier        Motor Diesel V8  2007   
12                                     S-18               Motor V6  2001   
13                                       J5               Motor V6  2019   
14                             F-Type Coupé  Motor 5.0 V8 Bi-Turbo  2009   
15                                      RAM        Motor Diesel V8  2016   
16                                  Série 1          Motor 3.0 32v  2012   
17                                       A3           Motor 1.0 8v  2019   
18                                     XC90           Motor 1.0 8v  2016   
19                           RS 3 Sportback          Motor 3.0 32v  2017   
20                             Pajero Sport           Motor 1.0 8v  2002   
21                                Série 1 M               Motor V8  2019   
22                      Lamborghini Obvious        Motor Diesel V6  1994   
23                    Effa Hafei Picape Baú           Motor Diesel  1991   
24                   Lamborghini Murciélago  Motor 5.0 V8 Bi-Turbo  2019   
25                                      Up!        Motor Diesel V6  2019   
26                       Fiesta Rocam Sedan        Motor 4.0 Turbo  2016   
27                          Sandero Stepway          Motor 3.0 32v  2019   
28                                 Cayman S          Motor 3.0 32v  2002   
29                                      V40        Motor 4.0 Turbo  1991   
30                           Classe SLS AMG           Motor 1.0 8v  1997   
31                                   Gol G4          Motor 2.0 16v  1995   
32                                    Linea           Motor 1.0 8v  2018   
33                                Ford F100          Motor 2.0 16v  1991   
34                                       J2               Motor V8  2014   
35                         Aston Martin DBS               Motor V8  1995   
36                               Pajero TR4        Motor 2.4 Turbo  2019   
37                               Polo Sedan               Motor V6  2019   
38                               California           Motor 1.0 8v  2009   
39                                      S60        Motor 2.4 Turbo  2019   
40                                   Tucson          Motor 2.0 16v  2013   
41                                     HB20               Motor V6  2003   
42                                   Optima          Motor 1.8 16v  2019   
43                                    Logan        Motor 4.0 Turbo  2002   
44                        Range Rover Vogue           Motor 1.0 8v  2019   
45                                   Sonata  Motor 5.0 V8 Bi-Turbo  2002   
46                    Ford New Fiesta Sedan          Motor 3.0 32v  2004   
47                     Série 5 Gran Turismo        Motor 4.0 Turbo  2001   
48                                     CR-V        Motor Diesel V8  2019   
49                            Ford EcoSport           Motor 1.0 8v  2003   
50                                    Versa          Motor 3.0 32v  2009   
51                                 Classe E           Motor 1.0 8v  2006   
52         

*É possível obtermos os tipos de dados que temos no conjunto, algo que também fizemos no curso anterior, utilizando o comando dataset.dtypes.*

In [ ]:
dataset.dtypes

Nome              object
Motor             object
Ano                int64
Quilometragem    float64
Zero_km             bool
Acessórios        object
Valor            float64
dtype: object

*Aqui temos os tipos de cada coluna do dataframe. **No Pandas, todo valor do tipo string é tratado como object.** Além deste, temos o ano como tipo inteiro (int64), quilometragem como float64, Zero_km como booleano (bool) e assim por diante. Note que, diferentemente dos arrays Numpy, nos dataframes do Pandas é possível trabalharmos com dados de tipos diferentes sem nenhum problema.*

*Vamos supor que desejamos calcular uma estatística descritiva com esse conjunto. Em nosso exemplo, usaremos as colunas "quilometragem" e "valor". Usaremos a variável dataset, que contém o nosso conjunto, e passaremos entre dois colchetes (pois estamos passando mais de uma informação) as duas variáveis que possuem os dados desejados - Quilometragem e Valor). Por fim, adicionaremos um ponto (.) e a função describe().*

In [ ]:
dataset[['Quilometragem', 'Valor']].describe()

Quilometragem          Valor
count     197.000000     258.000000
mean    58278.421320   98960.513101
std     35836.733259   29811.932305
min       107.000000   50742.100000
25%     27505.000000   70743.512500
50%     55083.000000   97724.380000
75%     90495.000000  124633.302500
max    119945.000000  149489.920000

*O Pandas gerará, de forma simples e rápida, um conjunto de estatísticas descritivas dessas duas variáveis: a quantidade de valores, a média, o desvio padrão, o valor mínimo, os quartis da distribuição e o valor máximo.*

*Com a função info() também podemos conseguir outras informações importantes. Por exemplo, acabamos de descobrir que temos menos ocorrências de Quilometragem do que de Valor.*

In [ ]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 258 entries, 0 to 257
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Nome           258 non-null    object 
 1   Motor          258 non-null    object 
 2   Ano            258 non-null    int64  
 3   Quilometragem  197 non-null    float64
 4   Zero_km        258 non-null    bool   
 5   Acessórios     258 non-null    object 
 6   Valor          258 non-null    float64
dtypes: bool(1), float64(2), int64(1), object(3)
memory usage: 12.5+ KB


*Isso acontece pois, dentro do conjunto, temos alguns valores faltando (missing) ou nulos (null). Nos aprofundaremos mais no tratamento desses valores na última aula do curso. No próximo capítulo voltaremos a trabalhar com o Python, agora conhecendo as tuplas*

# <font color=green> 2. TRABALHANDO COM TUPLAS
---

# 2.1.1 Criando tuplas

Tuplas são sequências imutáveis que são utilizadas para armazenar coleções de itens, geralmente heterogêneos. Podem ser construídas de várias formas:
```
- Utilizando um par de parênteses: ( )
- Utilizando uma vírgula à direita: x,
- Utilizando um par de parênteses com itens separados por vírgulas: ( x, y, z )
- Utilizando: tuple() ou tuple(iterador)
```

In [ ]:
nomes_carros = tuple(['Jetta Variant', 'Passat', 'Crossfox', 'DS5'])
nomes_carros

('Jetta Variant', 'Passat', 'Crossfox', 'DS5')

In [ ]:
nomes_carros[0]

'Jetta Variant'

In [ ]:
nomes_carros[-2]

'Crossfox'

In [ ]:
nomes_carros[2]

'Crossfox'

In [ ]:
nomes_carros[1:3]

('Passat', 'Crossfox')

In [ ]:
nomes_carros = ('Jetta Variant', 'Passat', 'Crossfox', 'DS5', ('Fusca', 'Gol', 'C4'))
nomes_carros

('Jetta Variant', 'Passat', 'Crossfox', 'DS5', ('Fusca', 'Gol', 'C4'))

In [ ]:
nomes_carros[-1][0]

'Fusca'

# 2.2 Seleções em tuplas

In [ ]:
carros = (
    ('Jetta Variant','Motor 4.0 Turbo', 2003, False, ('Rodas de liga', 'Travas elétricas', 'Piloto automático')),
    ('Passat', 'Motor Diesel', 1991, True, ('Central multimídia', 'Teto panorâmico', 'Freios ABS'))
    )

print(carros)


(('Jetta Variant', 'Motor 4.0 Turbo', 2003, False, ('Rodas de liga', 'Travas elétricas', 'Piloto automático')), ('Passat', 'Motor Diesel', 1991, True, ('Central multimídia', 'Teto panorâmico', 'Freios ABS')))


In [ ]:
carros[0][3]

False

In [ ]:
carros[-1][-1][-1]

'Freios ABS'

In [ ]:
carros[0][-1][:2]

('Rodas de liga', 'Travas elétricas')

# 2.3 Iterando em tuplas

In [ ]:
nomes_carros = ('Jetta Variant', 'Passat', 'Crossfox', 'DS5')
nomes_carros

('Jetta Variant', 'Passat', 'Crossfox', 'DS5')

*Para iterarmos por seus itens, usaremos a instrução for, atribuiremos um nome válido a cada item encontrado (por exemplo item) e continuaremos com in seguido do nome da tupla (nomes_carros). Não podemos nos esquecer de finalizar a linha com dois pontos (:) e da indentação na linha seguinte. Nesse exemplo simplesmente imprimiremos cada um dos item encontrados usando a função print().*

In [ ]:
for item in nomes_carros:
  print(item)

Jetta Variant
Passat
Crossfox
DS5


### Desempacotamento de tuplas

*Desempacotamento de tuplas é, basicamente, atribuirmos cada valor dentro de uma tupla a uma variável específica. Nesse primeiro exemplo, usaremos as variáveis carro_1, carro_2, carro_3 e carro_4 separadas por vírgula. Essa sequência receberá o conteúdo de nomes_carros*

*Esse é o desempacotamento de tuplas, algo útil, por exemplo, quando uma função nos retorna mais de um valor dentro de uma tupla e queremos utilizá-los de formas variadas no nosso programa, separando-os em variáveis diferentes.*

In [ ]:
carro_1, carro_2, carro_3, carro_4 = nomes_carros

In [ ]:
carro_1

In [ ]:
carro_4

'DS5'

*Prosseguiremos para outra situação interessante: imagine que uma função nos retorna quatro valores, mas queremos utilizar somente dois deles. Nesse caso, queremos somente o segundo e o último valor da tupla nomes_carros (a primeira posição e a terceira). Para isso, podemos usar o recurso do underscore (ou "underline"), da mesma forma que vimos nas listas.*

In [ ]:
_, A, _, B = nomes_carros

A

'Passat'

*Em outra situação, queremos somente um valor da nossa tupla, mas nenhum outro. Imagine se tivéssemos uma tupla com 60 itens e só gostaríamos de salvar o segundo. Com a sintaxe que acabamos de aprender, seria bastante trabalhoso recuperarmos esse valor.*

*Para simplificar, podemos ignorar o primeiro valor com underscore, receber o segundo com uma variável (por exemplo C) e finalizar com asterisco seguido de underscore ( *_). Dessa forma ignoraremos todos os valores posteriores.*

In [ ]:
_, C, *_ = nomes_carros

C

'Passat'

## *zip() built-in function*

https://docs.python.org/3.6/library/functions.html#zip

*Existe também uma nova função nativa do Python, chamada zip(), que cria um iterador com tuplas (de maneira semelhante à função range()), e vamos entender como ela funciona. Para nossos testes, inicializaremos duas listas: carros, contendo os modelos de alguns veículos; e valores, com seus preços.*

In [ ]:
carros = ['Jetta Variant', 'Passat', 'Crossfox', 'DS5']
carros

['Jetta Variant', 'Passat', 'Crossfox', 'DS5']

In [ ]:
valores = [88078.64, 106161.94, 72832.16, 124549.07]
valores

[88078.64, 106161.94, 72832.16, 124549.07]

In [ ]:
zip(carros, valores)

In [ ]:
list(zip(carros, valores))

[('Jetta Variant', 88078.64),
 ('Passat', 106161.94),
 ('Crossfox', 72832.16),
 ('DS5', 124549.07)]

*Repare que, interessantemente, nosso iterador pegou o primeiro elemento de cada lista e criou uma tupla; depois o segundo elemento de cada lista, criando uma segunda tupla; e assim sucessivamente até o final da lista. Isso nos traz a possibilidade de, por exemplo, iterar o nome e o valor de um veículo ao mesmo tempo.*

*Faremos isso usando a instrução for que irá iterar pelo resultado de zip(carros, valores). Por enquanto apenas imprimiremos as tuplas resultantes.*

In [ ]:
for item in zip(carros, valores):
  print(item)

('Jetta Variant', 88078.64)
('Passat', 106161.94)
('Crossfox', 72832.16)
('DS5', 124549.07)


*Aqui, ao invés de item, passaremos dois nomes separados por vírgula: carro e valor. Também passaremos a imprimir ambas as variáveis (lembrando que o print() aceita mais de um argumento).*

In [ ]:
for carro, valor in zip(carros, valores):
  print(carro, valor)

Jetta Variant 88078.64
Passat 106161.94
Crossfox 72832.16
DS5 124549.07


*Isso nos traz diversas outras possibilidades de análise. Por exemplo, podemos incluir uma condicional if que imprime somente os carros cujo valor é maior do que 100000.*

In [ ]:
for carro, valor in zip(carros, valores):
  if(valor > 100000):
    print(carro)

Passat
DS5


In [ ]:
nomes_carros = ('Jetta Variant', 'Passat', 'Crossfox', 'DS5')
nomes_carros

# <font color=green> 3. TRABALHANDO COM DICIONÁRIOS
---

# 3.1 Criando dicionários

Listas são coleções sequenciais, isto é, os itens destas sequências estão ordenados e utilizam índices (números inteiros) para acessar os valores.

Os dicionários são coleções um pouco diferentes. São estruturas de dados que representam um tipo de mapeamento. Mapeamentos são coleções de associações entre pares de valores onde o primeiro elemento do par é conhecido como chave (*key*) e o segundo como valor (*value*).

```
dicionario = {key_1: value_1, key_2: value_2, ..., key_n: value_n}
```

https://docs.python.org/3.6/library/stdtypes.html#typesmapping

In [ ]:
carros = ['Jetta Variant', 'Passat', 'Crossfox']
carros

['Jetta Variant', 'Passat', 'Crossfox']

In [ ]:
valores = [88078.64, 106161.94, 72832.16]
valores

[88078.64, 106161.94, 72832.16]

*Exibe a posição do objeto*

In [ ]:
carros.index('Passat')

1

*Exibe um objeto daquela lista, que estando na mesma posição que o objeto selecionado da outra lista.*

In [ ]:
valores[carros.index('Passat')]

106161.94

In [ ]:
dados = {'Jetta Variant': 88078.64, 'Passat':106161.94, 'Crossfox':72832.16  }
dados

{'Jetta Variant': 88078.64, 'Passat': 106161.94, 'Crossfox': 72832.16}

In [ ]:
type(dados)

dict

### Criando dicionários com *zip()*

In [ ]:
list(zip(carros, valores))

[('Jetta Variant', 88078.64), ('Passat', 106161.94), ('Crossfox', 72832.16)]

In [ ]:
dados = dict(zip(carros, valores))
dados

{'Jetta Variant': 88078.64, 'Passat': 106161.94, 'Crossfox': 72832.16}

Listas são estruturas de dados que representam um tipo básico de sequência.

Mapeamentos são coleções sequenciais ordenadas, que utilizam índices para acessar os valores

# 3.2 Operações com dicionários

## *dict[ key ]*

Retorna o valor correspondente à chave (*key*) no dicionário.

In [ ]:
dados = {'Jetta Variant': 88078.64, 'Passat': 106161.94, 'Crossfox': 72832.16}
dados

{'Jetta Variant': 88078.64, 'Passat': 106161.94, 'Crossfox': 72832.16}

In [ ]:
dados['Passat']

106161.94

## *key in dict*

Retorna **True** se a chave (*key*) for encontrada no dicionário.

In [ ]:
'Passat' in dados

True

In [ ]:
88078.64 in dados

False

In [ ]:
'Passat' not in dados

False

## *len(dict)*

Retorna o número de itens do dicionário.

In [ ]:
len(dados)

3

## *dict[ key ] = value*

Inclui um item ao dicionário.

In [ ]:
dados['DS5'] = 124549.07
dados

{'Jetta Variant': 88078.64,
 'Passat': 106161.94,
 'Crossfox': 72832.16,
 'DS5': 124549.07}

## *del dict[ key ]*

Remove o item de chave (*key*) do dicionário.

In [ ]:
del dados['Passat']
dados

{'Jetta Variant': 88078.64, 'Crossfox': 72832.16, 'DS5': 124549.07}

Operações básicas com dicionários

In [ ]:
dados = {
    'Passat': {
        'ano': 2012,
        'km': 50000,
        'valor': 75000,
        'acessorios': ['Airbag', 'ABS']
    },
    'Crossfox': {
        'ano': 2015,
        'km': 35000,
        'valor': 25000
    }
}

In [ ]:
'acessorios' in dados['Crossfox']

False

In [ ]:
'acessorios' in dados['Passat']

True

In [ ]:
dados['Crossfox']['valor']

25000

In [ ]:
dados['Passat']['acessorios'][-1]

'ABS'

# 3.3 Métodos de dicionários

## *dict.update()*

Atualiza o dicionário.

In [ ]:
dados = {'Jetta Variant': 88078.64, 'Passat':106161.94, 'Crossfox':72832.16  }

In [ ]:
dados

{'Jetta Variant': 88078.64, 'Passat': 106161.94, 'Crossfox': 72832.16}

In [ ]:
dados.update({'DS5': 124549.07})
dados

{'Jetta Variant': 88078.64,
 'Passat': 106161.94,
 'Crossfox': 72832.16,
 'DS5': 124549.07}

In [ ]:
dados.update({'DS5': 134323.95, 'Fusca': 124343.76})
dados

{'Jetta Variant': 88078.64,
 'Passat': 106161.94,
 'Crossfox': 72832.16,
 'DS5': 134323.95,
 'Fusca': 124343.76}

## *dict.copy()*

Cria uma cópia do dicionário.

In [ ]:
dadosCopy = dados.copy()

In [ ]:
dadosCopy

{'Jetta Variant': 88078.64,
 'Passat': 106161.94,
 'Crossfox': 72832.16,
 'DS5': 134323.95,
 'Fusca': 124343.76}

In [ ]:
del dadosCopy['Fusca']
dadosCopy

{'Jetta Variant': 88078.64,
 'Passat': 106161.94,
 'Crossfox': 72832.16,
 'DS5': 134323.95}

In [ ]:
dados

{'Jetta Variant': 88078.64,
 'Passat': 106161.94,
 'Crossfox': 72832.16,
 'DS5': 134323.95,
 'Fusca': 124343.76}

## *dict.pop(key[, default ])*

Se a chave for encontrada no dicionário, o item é removido e seu valor é retornado. Caso contrário, o valor especificado como *default* é retornado. Se o valor *default* não for fornecido e a chave não for encontrada no dicionário um erro será gerado.

In [ ]:
dadosCopy

{'Jetta Variant': 88078.64,
 'Passat': 106161.94,
 'Crossfox': 72832.16,
 'DS5': 134323.95}

In [ ]:
dadosCopy.pop('Passat')

106161.94

In [ ]:
dadosCopy

{'Jetta Variant': 88078.64, 'Crossfox': 72832.16, 'DS5': 134323.95}

Se o .pop não encontrar a chave ele gera um erro, então, um recurso é colocar uma vírgula seguida de um texto.

In [ ]:
dadosCopy.pop('Passat', 'Chave não encontrada')

'Chave não encontrada'

Mas se ele encontrar a chave ele vai executar o .pop

In [ ]:
dadosCopy.pop('DS5', 'Chave não encontrada')

134323.95

In [ ]:
dadosCopy

{'Jetta Variant': 88078.64, 'Crossfox': 72832.16}

## *dict.clear()*

Remove todos os itens do dicionário.

In [ ]:
dadosCopy.clear()

In [ ]:
dadosCopy

{}

# 3.4 Iterando em dicionários

In [ ]:
dados = {'Crossfox': 72832.16, 'DS5': 124549.07,  'Fusca': 150000,  'Jetta Variant': 88078.64,  'Passat': 106161.95}
dados

{'Crossfox': 72832.16,
 'DS5': 124549.07,
 'Fusca': 150000,
 'Jetta Variant': 88078.64,
 'Passat': 106161.95}

## *dict.keys()*

Retorna uma lista contendo as chaves (*keys*) do dicionário.

In [ ]:
dados.keys()

dict_keys(['Crossfox', 'DS5', 'Fusca', 'Jetta Variant', 'Passat'])

Podemos usar essa instrução dentro de um for(), como se fosse um iterador. Nesse caso, chamaremos cada elemento encontrado de key (ou seja, cada chave) e imprimiremos dados[key], o que nos retornará o valor daquela chave.

In [ ]:
for key in dados.keys():
  print(dados[key])

72832.16
124549.07
150000
88078.64
106161.95


## *dict.values()*

O método values() funciona praticamente da mesma forma, mas nos retorna os valores contidos naquele dicionário.

In [ ]:
dados.values()

dict_values([72832.16, 124549.07, 150000, 88078.64, 106161.95])

## *dict.items()*

Retorna uma lista contendo uma tupla para cada par chave-valor (*key-value*) do dicionário.

In [ ]:
dados.items()

dict_items([('Crossfox', 72832.16), ('DS5', 124549.07), ('Fusca', 150000), ('Jetta Variant', 88078.64), ('Passat', 106161.95)])

In [ ]:
for item in dados.items():
    print(item)

('Crossfox', 72832.16)
('DS5', 124549.07)
('Fusca', 150000)
('Jetta Variant', 88078.64)
('Passat', 106161.95)


In [ ]:
for key, value in dados.items():
    print(key, value)

Crossfox 72832.16
DS5 124549.07
Fusca 150000
Jetta Variant 88078.64
Passat 106161.95


In [ ]:
for key, value in dados.items():
  if (value > 100000):
    print(key)

DS5
Fusca
Passat


# <font color=green> 4. FUNÇÕES E PACOTES
---
    
Funções são unidades de código reutilizáveis que realizam uma tarefa específica, podem receber alguma entrada e também podem retornar alguma resultado.

# 4.1 Built-in function

A linguagem Python possui várias funções integradas que estão sempre acessíveis. Algumas já utilizamos em nosso treinamento: type(), print(), zip(), len(), set() etc.

https://docs.python.org/3.6/library/functions.html

In [ ]:
dados = {'Jetta Variant': 88078.64, 'Passat': 106161.94, 'Crossfox': 72832.16}
dados

{'Jetta Variant': 88078.64, 'Passat': 106161.94, 'Crossfox': 72832.16}

Queremos obter a somatória dos valores dos veículos desse conjunto, algo que pode ser obtido de algumas formas mais complexas e de outras mais simples.

1. Criamos uma lista vazia valores e iniciaremos um iterador for chamando cada item encontrado em dados.values() de valor e adicionando tais valores na lista por meio da função append(). Desse modo obteremos todos os valores dentro de uma única lista, o que nos permitirá somá-los.

In [ ]:
valores = []

for valor in dados.values():
  valores.append(valor)
valores

[88078.64, 106161.94, 72832.16]

2. Outra forma de obtermos a somatória é criando uma variável soma que se iniciará com 0, iterando pelos valores com dados.values() e, ao invés de adicioná-los a uma lista, somá-los diretamente à variável criada anteriormente com soma += valor.

In [ ]:
soma = 0
for valor in dados.values():
  soma += valor
soma

267072.74

Dentre as built-in functions do Python, existe uma que pode tornar nosso código ainda mais simples: sum(), que é justamente "soma" em inglês. Na documentação da função sum() verificaremos que a função pode receber dois parâmetros: um iterador iterable e um start. Este último aparece entre colchetes, o que significa duas coisas: que o parâmetro possui um valor padrão e que ele é dispensável - ou seja, que não é necessário fornecê-lo à função para que ela nos retorne um valor.

Antes de utilizarmos essa função, repare na devolução da chamada list(dados.values()):

In [ ]:
list(dados.values())

[88078.64, 106161.94, 72832.16]

Aqui conseguimos obter uma lista com todos os valores do nosso dicionário, da mesma maneira que anteriormente, mas utilizando um código bem mais simples e sucinto. A soma (sum()) operará da mesma forma, bastando passarmos como parâmetro o nosso dados.values()

In [ ]:
sum(dados.values())

267072.74

Nosso resultado é a soma de todos os valores do dicionário dados, executada de maneira bastante resumida. O parâmetro start dessa função significa que podemos iniciar a soma já com algum valor. Isso pode ser feito simplesmente passando o valor desejado após uma vírgula.

Assim, a nossa soma será adicionada ao valor inicial - nesse caso, um milhão.



In [ ]:
sum(dados.values(), 1000000)

1267072.7399999998

*Nota*
Quando precisamos entender determinadas funcionalidades, é interessante consultarmos a documentação. O Python possui também uma função nativa help() para qual podemos passar, como parâmetro, a função sobre a qual queremos mais informações - por exemplo, print.

Isso nos devolverá a "ajuda" da função, que é uma espécie de documentação resumida.

In [ ]:
help(print)

Help on built-in function print in module builtins:

print(...)
    print(value, ..., sep=' ', end='\n', file=sys.stdout, flush=False)
    
    Prints the values to a stream, or to sys.stdout by default.
    Optional keyword arguments:
    file:  a file-like object (stream); defaults to the current sys.stdout.
    sep:   string inserted between values, default a space.
    end:   string appended after the last value, default a newline.
    flush: whether to forcibly flush the stream.



Outra forma de acessarmos esse conteúdo no Colab é passarmos, depois do nome da função, uma interrogação.Isso abrirá uma aba do próprio Colab com as informações da função.



In [ ]:
print?

# 4.2 Definindo funções sem e com parâmetros

Utilizaremos a palavra chave def seguida do nome da função e da abertura e fechamento de parênteses (que ficarão vazios, já que a função não receberá parâmetros) e, por fim, dois pontos. Na linha seguinte, precedidas por uma indentação, teremos as instruções que a função executará.

### Funções sem parâmetros

#### Formato padrão

```
def <nome>():
    <instruções>
```

Como queremos manter o código simples, iremos simplesmente calcular uma média. Nas instruções, criaremos uma variável valor para qual atribuiremos o cálculo (1 + 2 + 3) / 3 - ou seja, a soma de três valores e a sua divisão pelo número de valores. Por fim, imprimiremos o resultado guardado em valor.

Para testarmos, executaremos a nova função com media().

In [ ]:
def media():
    valor = (1 + 2 + 3) / 3
    print(valor)

media()

2.0


### Funções com parâmetros

#### Formato padrão

```
def <nome>(<param_1>, <param_2>, ..., <param_n>):
    <instruções>
```

In [ ]:
def media(number_1, number_2, number_3):
  valor = (number_1 + number_2 +number_3) / 3
  print(valor)

In [ ]:
media(1,2,3)

2.0


A diferença é que agora podemos passar qualquer outro valor, como 23, 45 e 67.

In [ ]:
media(23, 45, 67)

45.0


Outro ponto interessante é que, ao invés de passarmos esses valores sequencias, podemos utilizar uma lista. Criaremos outra função media(), que dessa vez recebe como parâmetro uma lista. Nas instruções, faremos o cálculo da média com o auxílio das built-in functions do Python: usaremos a função sum() para somarmos os itens da lista e dividiremos o resultado por len(lista), ou seja, a quantidade de valores que foram somados. Por fim, imprimiremos o valor.

In [ ]:
def media(lista):
  valor = sum(lista) / len(lista)
  print(valor)

In [ ]:
media ([1, 2, 3, 4, 5, 6, 7, 8, 9])

5.0


Uma coisa importante é que nossa função está imprimindo um valor, mas no momento não é possível atribuirmos tal valor a uma variável, afinal ele não é retornado. Por exemplo, se atribuirmos a chamada de media() a uma variável resultado, teremos a impressão do nosso valor normalmente.

In [ ]:
resultado = media ([1, 2, 3, 4, 5, 6, 7, 8, 9])
resultado

5.0


Ao verificarmos o tipo de resultado, receberemos um "NoneType".Ou seja, a variável na verdade não tem valor nenhum.

In [ ]:
type(resultado)

NoneType

# 4.3 Definindo funções que retornam valores

Construímos uma função que não retorna valores e, por isso, não conseguimos atribuir o seu resultado a uma variável para utilizá-lo no código posteriormente. O formato padrão de uma função que retorna um valor não é muito diferente daquele que já estamos acostumados.

### Funções que retornam um valor

#### Formato padrão

```
def <nome>(<param_1>, <param_2>, ..., <param_n>):
    <instruções>
    return <resultado>
```

### Funções que retornam mais de um valor

#### Formato padrão

```
def <nome>(<param_1>, <param_2>, ..., <param_n>):
    <instruções>
    return (<resultado_1>, <resultado_2>, ..., <resultado_n>)
```

# <font color=green> 5. PANDAS BÁSICO
---

**versão: 0.25.2**
  
Pandas é uma ferramenta de manipulação de dados de alto nível, construída com base no pacote Numpy. O pacote pandas possui estruturas de dados bastante interessantes para manipulação de dados e por isso é muito utilizado por cientistas de dados.


## Estruturas de Dados

### Series

Series são arrays unidimensionais rotulados capazes de armazenar qualquer tipo de dado. Os rótulos das linhas são chamados de **index**. A forma básica de criação de uma Series é a seguinte:


```
    s = pd.Series(dados, index = index)
```

O argumento *dados* pode ser um dicionário, uma lista, um array Numpy ou uma constante.

### DataFrames

DataFrame é uma estrutura de dados tabular bidimensional com rótulos nas linha e colunas. Como a Series, os DataFrames são capazes de armazenar qualquer tipo de dados.


```
    df = pd.DataFrame(dados, index = index, columns = columns)
```

O argumento *dados* pode ser um dicionário, uma lista, um array Numpy, uma Series e outro DataFrame.

**Documentação:** https://pandas.pydata.org/pandas-docs/version/0.25/

# 5.1 Estruturas de dados

In [ ]:
import pandas as pd

### Criando uma Series a partir de uma lista

In [ ]:
carros = ['Jetta Variant', 'Passat', 'Crossfox']
carros

['Jetta Variant', 'Passat', 'Crossfox']

In [ ]:
pd.Series(carros)

0    Jetta Variant
1           Passat
2         Crossfox
dtype: object

### Criando um DataFrame a partir de uma lista de dicionários

Para nosso exemplo, inicializaremos a variável dados como uma lista na qual cada item é um dicionário contendo as informações de um veículo.

In [ ]:
dados = [
    {'Nome': 'Jetta Variant', 'Motor': 'Motor 4.0 Turbo', 'Ano': 2003, 'Quilometragem': 44410.0, 'Zero_km': False, 'Valor': 88078.64},
    {'Nome': 'Passat', 'Motor': 'Motor Diesel', 'Ano': 1991, 'Quilometragem': 5712.0, 'Zero_km': False, 'Valor': 106161.94},
    {'Nome': 'Crossfox', 'Motor': 'Motor Diesel V8', 'Ano': 1990, 'Quilometragem': 37123.0, 'Zero_km': False, 'Valor': 72832.16}
]

Criaremos então uma variável dataset que receberá a chamada de pd.DataFrame() passando como parâmetro nossa lista dados. Note que em DataFrame() ambas as consoantes "D" e "F" devem ser maiúsculas, do contrário receberemos um erro.

In [ ]:
dataset = pd.DataFrame(dados)
dataset

Nome            Motor   Ano  Quilometragem  Zero_km      Valor
0  Jetta Variant  Motor 4.0 Turbo  2003        44410.0    False   88078.64
1         Passat     Motor Diesel  1991         5712.0    False  106161.94
2       Crossfox  Motor Diesel V8  1990        37123.0    False   72832.16

Se quisermos modificar a ordem das colunas desse dataframe, basta passarmos o seu nome (dataset) seguido da abertura de dois pares de colchetes ([[]]) dentro dos quais passaremos nossos rótulos na ordem desejada.

In [ ]:
dataset[['Nome', 'Quilometragem', 'Motor', 'Ano', 'Zero_km', 'Valor']]

Nome  Quilometragem            Motor   Ano  Zero_km      Valor
0  Jetta Variant        44410.0  Motor 4.0 Turbo  2003    False   88078.64
1         Passat         5712.0     Motor Diesel  1991    False  106161.94
2       Crossfox        37123.0  Motor Diesel V8  1990    False   72832.16

### Criando um DataFrame a partir de um dicionário

In [ ]:
dados = {
    'Nome': ['Jetta Variant', 'Passat', 'Crossfox'],
    'Motor': ['Motor 4.0 Turbo', 'Motor Diesel', 'Motor Diesel V8'],
    'Ano': [2003, 1991, 1990],
    'Quilometragem': [44410.0, 5712.0, 37123.0],
    'Zero_km': [False, False, False],
    'Valor': [88078.64, 106161.94, 72832.16]
}

Agora estamos trabalhando com um dicionário no qual as chaves são os rótulos dos dados e os valores são uma lista com todas as informações, como nomes, motores, anos e assim por diante, todos seguindo a mesma ordem. Executando novamente pd.DataFrame(dados), criaremos nosso novo dataset.



In [ ]:
dataset = pd.DataFrame(dados)
dataset

Nome            Motor   Ano  Quilometragem  Zero_km      Valor
0  Jetta Variant  Motor 4.0 Turbo  2003        44410.0    False   88078.64
1         Passat     Motor Diesel  1991         5712.0    False  106161.94
2       Crossfox  Motor Diesel V8  1990        37123.0    False   72832.16

### Criando um DataFrame a partir de uma arquivo externo

Usaremos esse arquivo para criarmos um dataframe a partir de dados externos. Na realidade, essa é a forma mais comum de criar dataframes.

Feito o upload, usaremos a função pd.read_csv() para lermos o arquivo. Lembre-se que devemos passar como argumento, além do nome do arquivo (db.csv), um parâmetro sep que define o separador utilizado nesse conjunto - no nosso caso, ;.

Se você abrir o arquivo db.csv em um editor de texto, poderá verificar que os dados realmente estão separados por ponto e vírgula.

In [ ]:
dataset = pd.read_csv('db.csv', sep = ';')
dataset

Nome                  Motor   Ano  Quilometragem  \
0            Jetta Variant        Motor 4.0 Turbo  2003        44410.0   
1                   Passat           Motor Diesel  1991         5712.0   
2                 Crossfox        Motor Diesel V8  1990        37123.0   
3                      DS5        Motor 2.4 Turbo  2019            NaN   
4         Aston Martin DB4        Motor 2.4 Turbo  2006        25757.0   
..                     ...                    ...   ...            ...   
253           Phantom 2013               Motor V8  2014        27505.0   
254  Cadillac Ciel concept               Motor V8  1991        29981.0   
255             Classe GLK  Motor 5.0 V8 Bi-Turbo  2002        52637.0   
256       Aston Martin DB5           Motor Diesel  1996         7685.0   
257                  Macan        Motor Diesel V6  1992        50188.0   

     Zero_km                                         Acessórios      Valor  
0      False  ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
1      False  ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
2      False  ['Piloto automático', 'Controle de estabilidad...   72832.16  
3       True  ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
4      False  ['Rodas de liga', '4 X 4', 'Central multimídia...   92612.10  
..       ...                                                ...        ...  
253    False  ['Controle de estabilidade', 'Piloto automátic...   51759.58  
254    False  ['Bancos de couro', 'Painel digital', 'Sensor ...   51667.06  
255    False  ['Rodas de liga', 'Controle de tração', 'Câmbi...   68934.03  
256    False  ['Ar condicionado', '4 X 4', 'Câmbio automátic...  122110.90  
257    False  ['Central multimídia', 'Teto panorâmico', 'Vid...   90381.47  

[258 rows x 7 columns]

o importarmos um arquivo com a função pd.read_csv(), também podemos passar um parâmetro index_col = 0, definindo a coluna que desejamos assumir como índice do dataframe.

Note que, ao invés do índice numérico, passaremos a assumir a variável Nome como índice.

In [ ]:
dataset = pd.read_csv('db.csv', sep = ';', index_col = 0)
dataset

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                                         
Jetta Variant                Motor 4.0 Turbo  2003        44410.0    False   
Passat                          Motor Diesel  1991         5712.0    False   
Crossfox                     Motor Diesel V8  1990        37123.0    False   
DS5                          Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4             Motor 2.4 Turbo  2006        25757.0    False   
...                                      ...   ...            ...      ...   
Phantom 2013                        Motor V8  2014        27505.0    False   
Cadillac Ciel concept               Motor V8  1991        29981.0    False   
Classe GLK             Motor 5.0 V8 Bi-Turbo  2002        52637.0    False   
Aston Martin DB5                Motor Diesel  1996         7685.0    False   
Macan                        Motor Diesel V6  1992        50188.0    False   

                                                              Acessórios  \
Nome                                                                       
Jetta Variant          ['Rodas de liga', 'Travas elétricas', 'Piloto ...   
Passat                 ['Central multimídia', 'Teto panorâmico', 'Fre...   
Crossfox               ['Piloto automático', 'Controle de estabilidad...   
DS5                    ['Travas elétricas', '4 X 4', 'Vidros elétrico...   
Aston Martin DB4       ['Rodas de liga', '4 X 4', 'Central multimídia...   
...                                                                  ...   
Phantom 2013           ['Controle de estabilidade', 'Piloto automátic...   
Cadillac Ciel concept  ['Bancos de couro', 'Painel digital', 'Sensor ...   
Classe GLK             ['Rodas de liga', 'Controle de tração', 'Câmbi...   
Aston Martin DB5       ['Ar condicionado', '4 X 4', 'Câmbio automátic...   
Macan                  ['Central multimídia', 'Teto panorâmico', 'Vid...   

                           Valor  
Nome                              
Jetta Variant           88078.64  
Passat                 106161.94  
Crossfox                72832.16  
DS5                    124549.07  
Aston Martin DB4        92612.10  
...                          ...  
Phantom 2013            51759.58  
Cadillac Ciel concept   51667.06  
Classe GLK              68934.03  
Aston Martin DB5       122110.90  
Macan                   90381.47  

[258 rows x 6 columns]

# 5.2 Seleções com DataFrames

Começaremos pela função head(), que mostra os cinco primeiros registros do nosso conjunto de dados.

In [ ]:
dataset.head()

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                              
Jetta Variant     Motor 4.0 Turbo  2003        44410.0    False   
Passat               Motor Diesel  1991         5712.0    False   
Crossfox          Motor Diesel V8  1990        37123.0    False   
DS5               Motor 2.4 Turbo  2019            NaN     True   
Aston Martin DB4  Motor 2.4 Turbo  2006        25757.0    False   

                                                         Acessórios      Valor  
Nome                                                                            
Jetta Variant     ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
Passat            ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
Crossfox          ['Piloto automático', 'Controle de estabilidad...   72832.16  
DS5               ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07  
Aston Martin DB4  ['Rodas de liga', '4 X 4', 'Central multimídia...   92612.10

### Selecionando colunas

Se quisermos selecionar somente a coluna "Valor", por exemplo, bastará passarmos o seu rótulo entre colchetes.

In [ ]:
dataset['Valor']

Nome
Jetta Variant             88078.64
Passat                   106161.94
Crossfox                  72832.16
DS5                      124549.07
Aston Martin DB4          92612.10
                           ...    
Phantom 2013              51759.58
Cadillac Ciel concept     51667.06
Classe GLK                68934.03
Aston Martin DB5         122110.90
Macan                     90381.47
Name: Valor, Length: 258, dtype: float64

Note que também recebemos os nomes dos veículos. Isso acontece pois dataset['Valor'] nos retorna, além da coluna desejada, os índices, e anteriormente definimos que o rótulo "Nome" deveria ser usado como índice. Também deve-se prestar atenção ao fato de que esse dado é uma series do Pandas, algo que podemos conferir com a função type().

In [ ]:
type(dataset['Valor'])

pandas.core.series.Series

Ou seja, um dataframe não deixa de ser um conjunto de series. Se quisermos, também é possível selecionar uma única coluna e obter como retorno um dataframe. Para isso, ao invés de passarmos um único par de colchetes, passaremos dois.

In [ ]:
dataset[['Valor']]

Valor
Nome                            
Jetta Variant           88078.64
Passat                 106161.94
Crossfox                72832.16
DS5                    124549.07
Aston Martin DB4        92612.10
...                          ...
Phantom 2013            51759.58
Cadillac Ciel concept   51667.06
Classe GLK              68934.03
Aston Martin DB5       122110.90
Macan                   90381.47

[258 rows x 1 columns]

### Selecionando linhas - [ i : j ]

<font color=red>**Observação:**</font> A indexação tem origem no zero e nos fatiamentos (*slices*) a linha com índice i é **incluída** e a linha com índice j **não é incluída** no resultado.

In [ ]:
dataset[0:3]

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                           
Jetta Variant  Motor 4.0 Turbo  2003        44410.0    False   
Passat            Motor Diesel  1991         5712.0    False   
Crossfox       Motor Diesel V8  1990        37123.0    False   

                                                      Acessórios      Valor  
Nome                                                                         
Jetta Variant  ['Rodas de liga', 'Travas elétricas', 'Piloto ...   88078.64  
Passat         ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
Crossfox       ['Piloto automático', 'Controle de estabilidad...   72832.16

Note que, como afirmado na observação, o primeiro índice é incluído e o último não. Existem outras formas mais elegantes de fazermos seleções, conhecidas como .loc e .iloc.



### Utilizando .loc para seleções

<font color=red>**Observação:**</font> Seleciona um grupo de linhas e colunas segundo os rótulos ou uma matriz booleana.

In [ ]:
dataset.loc['Passat']

Motor                                                 Motor Diesel
Ano                                                           1991
Quilometragem                                               5712.0
Zero_km                                                      False
Acessórios       ['Central multimídia', 'Teto panorâmico', 'Fre...
Valor                                                    106161.94
Name: Passat, dtype: object

Se quisermos mais de uma informação, precisaremos passá-las dentro de dois pares de colchetes. Usaremos como exemplo os veículos "Passat" e "DS5".

In [ ]:
dataset.loc[['Passat', 'DS5']]

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                    
Passat     Motor Diesel  1991         5712.0    False   
DS5     Motor 2.4 Turbo  2019            NaN     True   

                                               Acessórios      Valor  
Nome                                                                  
Passat  ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
DS5     ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07

O .loc também nos permite selecionar quais colunas desejamos exibir. Para isso, depois das informações das linhas, adicionaremos uma vírgula e um novo par de colchetes, dessa vez com as informações "Motor" e "Valor" representando as colunas a serem exibidas.

In [ ]:
dataset.loc[['Passat', 'DS5'], ['Motor', 'Valor']]

Motor      Valor
Nome                              
Passat     Motor Diesel  106161.94
DS5     Motor 2.4 Turbo  124549.07

Já se passarmos apenas : como primeiro parâmetro, receberemos todas as linhas do conjunto e somente as colunas selecionadas.

In [ ]:
dataset.loc[:, ['Motor', 'Valor']]

Motor      Valor
Nome                                                   
Jetta Variant                Motor 4.0 Turbo   88078.64
Passat                          Motor Diesel  106161.94
Crossfox                     Motor Diesel V8   72832.16
DS5                          Motor 2.4 Turbo  124549.07
Aston Martin DB4             Motor 2.4 Turbo   92612.10
...                                      ...        ...
Phantom 2013                        Motor V8   51759.58
Cadillac Ciel concept               Motor V8   51667.06
Classe GLK             Motor 5.0 V8 Bi-Turbo   68934.03
Aston Martin DB5                Motor Diesel  122110.90
Macan                        Motor Diesel V6   90381.47

[258 rows x 2 columns]

### Utilizando .iloc para seleções

<font color=red>**Observação:**</font> Seleciona com base nos índices, ou seja, se baseia na posição das informações.

In [ ]:
dataset.iloc[1]

Motor                                                 Motor Diesel
Ano                                                           1991
Quilometragem                                               5712.0
Zero_km                                                      False
Acessórios       ['Central multimídia', 'Teto panorâmico', 'Fre...
Valor                                                    106161.94
Name: Passat, dtype: object

Note que nos foram devolvidas as informações do carro de índice 1, que é o "Passat". Utilizando dois pares de colchetes ao invés de um, receberemos essas informações no formato de um dataframe.

In [ ]:
dataset.iloc[[1]]

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                 
Passat  Motor Diesel  1991         5712.0    False   

                                               Acessórios      Valor  
Nome                                                                  
Passat  ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94

Quando queremos informações de múltiplas linhas em um dataset, utilizaremos somente um par de colchetes, dentro dos quais passaremos o índice inicial e o final (lembrando que o inicial é inclusivo e o final não).

In [ ]:
dataset.iloc[1:4]

Motor   Ano  Quilometragem  Zero_km  \
Nome                                                      
Passat       Motor Diesel  1991         5712.0    False   
Crossfox  Motor Diesel V8  1990        37123.0    False   
DS5       Motor 2.4 Turbo  2019            NaN     True   

                                                 Acessórios      Valor  
Nome                                                                    
Passat    ['Central multimídia', 'Teto panorâmico', 'Fre...  106161.94  
Crossfox  ['Piloto automático', 'Controle de estabilidad...   72832.16  
DS5       ['Travas elétricas', '4 X 4', 'Vidros elétrico...  124549.07

Para obtermos colunas específicas, passaremos, após uma vírgula, um novo par de colchetes contendo os índices das colunas que desejamos acessar. Não é necessário passá-los na mesma ordem do dataset original, algo que exemplificaremos na seleção abaixo, feita com as colunas 0, 5 e 2 ("Motor", "Valor" e "Quilometragem", respectivamente).



In [ ]:
dataset.iloc[1:4, [0, 5, 2]]

Motor      Valor  Quilometragem
Nome                                               
Passat       Motor Diesel  106161.94         5712.0
Crossfox  Motor Diesel V8   72832.16        37123.0
DS5       Motor 2.4 Turbo  124549.07            NaN

Da mesma forma, é possível selecionar linhas específicas que não estejam em sequência, bastando passarmos seus índices dentro de um par de colchetes.

In [ ]:
dataset.iloc[[1, 42, 22], [0, 5, 2]]

Motor      Valor  Quilometragem
Nome                                                          
Passat                  Motor Diesel  106161.94         5712.0
Optima                 Motor 1.8 16v   86641.34            NaN
Lamborghini Obvious  Motor Diesel V6  133529.84        98079.0

Se quisermos exibir todas as linhas, usaremos os dois pontos (:).



In [ ]:
dataset.iloc[:, [0, 5, 2]]

Motor      Valor  Quilometragem
Nome                                                                  
Jetta Variant                Motor 4.0 Turbo   88078.64        44410.0
Passat                          Motor Diesel  106161.94         5712.0
Crossfox                     Motor Diesel V8   72832.16        37123.0
DS5                          Motor 2.4 Turbo  124549.07            NaN
Aston Martin DB4             Motor 2.4 Turbo   92612.10        25757.0
...                                      ...        ...            ...
Phantom 2013                        Motor V8   51759.58        27505.0
Cadillac Ciel concept               Motor V8   51667.06        29981.0
Classe GLK             Motor 5.0 V8 Bi-Turbo   68934.03        52637.0
Aston Martin DB5                Motor Diesel  122110.90         7685.0
Macan                        Motor Diesel V6   90381.47        50188.0

[258 rows x 3 columns]

# 5.3 Queries com DataFrames

### Utilizando o método query

# 5.4 Iterando com DataFrames

# 5.5 Tratamento de dados